prune.go

```diff
@@ -129,122 +135,194 @@ func txnsPruneC(txnsName string) string {
-		Id  bson.ObjectId `bson:"_id"`
-		Ops []txn.Op      `bson:"o"`
+	// Load the ids of all completed and aborted txns into a separate
+	// temporary collection.
+	logger.Debugf("loading all completed transactions")
+	matchCompleted := bson.M{"$match": bson.M{"s": bson.M{"$in": []int{taborted, tapplied}}}}
+	pipe := txns.Pipe([]bson.M{
+		matchCompleted,
+		{"$project": bson.M{"_id": 1}},
+		{"$out": workingSetName},
+	})
+	var dummy bson.D
+	if err := runPipe(pipe, &dummy); err != nil {
+		return fmt.Errorf("reading completed txns: %v", err)
 	}
 
-	logger.Debugf("loading all completed transactions")
-	completed := bson.M{
-		"s": bson.M{"$in": []int{taborted, tapplied}},
-	}
-	iter := txns.Find(completed).Select(bson.M{"_id": 1, "o": 1}).Iter()
-	for iter.Next(&txnDoc) {
-		txnIds[txnDoc.Id] = present
-		for _, op := range txnDoc.Ops {
-			collNames[op.C] = present
-		}
+	// Load all collections which have completed transactions.
+	logger.Debugf("loading collections with transactions")
+	pipe = txns.Pipe([]bson.M{
+		matchCompleted,
+		{"$project": bson.M{"_id": 0, "c": "$o.c"}},
+		{"$unwind": "$c"},
+		{"$group": bson.M{"_id": "$c"}},
+	})
+	var collNameDocs []bson.D
+	if err := runPipe(pipe, &collNameDocs); err != nil {
+		return fmt.Errorf("reading collection names: %v", err)
 	}
-	if err := iter.Close(); err != nil {
-		return fmt.Errorf("failed to read all txns: %v", err)
+	var collNames []string
+	for _, doc := range collNameDocs {
+		if len(doc) == 1 {
+			if collName, ok := doc[0].Value.(string); ok {
+				collNames = append(collNames, collName)
+			}
+		}
 	}
-	logger.Debugf("found %d completed transactions across %d collections",
-		len(txnIds), len(collNames))
+
+	logger.Debugf("%d collections with txns discovered", len(collNames))
 
 	// Transactions may also be referenced in the stash.
-	collNames["txns.stash"] = present
+	collNames = append(collNames, "txns.stash")
 
-	// Now remove the txn ids referenced by all documents in all
-	// txn using collections from the set of known txn ids.
+	// Now remove the txn ids referenced by any document in any
+	// txn-using collection from the set of known txn ids.
 	//
 	// Working the other way - starting with the set of txns
 	// referenced by documents and then removing any not in that set
 	// from the txns collection - is unsafe as it will result in the
-	// removal of transactions run while pruning executes.
-	//
-	for collName := range collNames {
-		logger.Tracef("checking %s for transaction references", collName)
+	// removal of transactions created during the pruning process.
+	t := newSimpleTimer(logInterval)
+	remover := newBulkRemover(workingSet)
+	defer remover.close()
+	for _, collName := range collNames {
+		logger.Tracef("checking %s for txn references", collName)
 		coll := db.C(collName)
 		var tDoc struct {
 			Queue []string `bson:"txn-queue"`
 		}
-		iter := coll.Find(nil).Select(bson.M{"txn-queue": 1}).Iter()
+		query := coll.Find(nil).Select(bson.M{"txn-queue": 1})
```

Comment:
I assume this is the equivalent of:
db.collection.find({}, {"txn-queue": 1})

I wonder if it would be helpful to do the same trick we did earlier with:
db.collection.find({}, {"_id": 0, "txn-queue": 1})

It also *really* feels like it wants to be another pipeline. Would it be sane to do the aggregation into another temp table so that we wouldn't have to iterate the docs?

Something like:
```
pipe = txns.Pipe([]bson.M{
		{"$project": bson.M{"_id": 0, "txn": "txn-queue"}},
		{"$unwind": "$txn"},
		{"$group": bson.M{"_id": "$txn"}},
	})
```

Feels really good here, as we don't end up double handling the same txn ids over and over again.